pkgs/applications/emulators/ppsspp/default.nix

```diff
@@ -1,72 +1,128 @@
-{ mkDerivation
+{ lib
+, stdenv
+, makeDesktopItem
+, makeWrapper
 , fetchFromGitHub
 , SDL2
 , cmake
 , ffmpeg
 , glew
-, lib
+, libffi
 , libzip
 , pkg-config
 , python3
-, qtbase
-, qtmultimedia
+, qtbase ? null
+, qtmultimedia ? null
 , snappy
+, vulkan-loader
+, wayland
+, wrapQtAppsHook ? null
 , zlib
+
+# Currently only enabled on SDL X11 builds
+, enableVulkan ? true
+
+# Headless build can only be enabled with SDL
+, enableHeadless ? true
+
+# Only SDL front end needs to specify whether to use Wayland
+, forceWayland ? false
 }:
 
-mkDerivation rec {
-  pname = "ppsspp";
+let
+  enableQt = (qtbase != null);
+  frontend = if enableQt then "Qt" else "SDL";
+  vulkanPath = lib.makeLibraryPath [ vulkan-loader ];
+
+  # experimental, see https://github.com/hrydgard/ppsspp/issues/13845
+  vulkanWayland = enableVulkan && forceWayland;
+in
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "ppsspp"
+    + lib.optionalString enableQt "-qt"
+    + lib.optionalString !(enableQt||enableHeadless) "-sdl"
+    + lib.optionalString forceWayland "-wayland";
   version = "1.13.1";
 
   src = fetchFromGitHub {
     owner = "hrydgard";
-    repo = pname;
-    rev = "v${version}";
+    repo = "ppsspp";
+    rev = "v${finalAttrs.version}";
     fetchSubmodules = true;
     sha256 = "sha256-WsFy2aSOmkII2Lte5et4W6qj0AXUKWWkYe88T0OQP08=";
   };
 
   postPatch = ''
-    substituteInPlace git-version.cmake --replace unknown ${src.rev}
+    substituteInPlace git-version.cmake --replace unknown ${finalAttrs.src.rev}
     substituteInPlace UI/NativeApp.cpp --replace /usr/share $out/share
   '';
 
-  nativeBuildInputs = [ cmake pkg-config python3 ];
+  nativeBuildInputs = [
+    cmake
+    makeWrapper
+    pkg-config
+    python3
+    wrapQtAppsHook
+  ];
 
   buildInputs = [
     SDL2
     ffmpeg
-    glew
+    (glew.override { enableEGL = forceWayland; })
     libzip
     qtbase
     qtmultimedia
     snappy
     zlib
-  ];
+  ] ++ lib.optional enableVulkan vulkan-loader
+    ++ lib.optionals vulkanWayland [ wayland libffi ];
 
   cmakeFlags = [
-    "-DHEADLESS=OFF"
+    "-DHEADLESS=${if enableHeadless then "ON" else "OFF"}"
     "-DOpenGL_GL_PREFERENCE=GLVND"
     "-DUSE_SYSTEM_FFMPEG=ON"
     "-DUSE_SYSTEM_LIBZIP=ON"
     "-DUSE_SYSTEM_SNAPPY=ON"
-    "-DUSING_QT_UI=ON"
+    "-DUSE_WAYLAND_WSI=${if vulkanWayland then "ON" else "OFF"}"
+    "-DUSING_QT_UI=${if enableQt then "ON" else "OFF"}"
   ];
 
+  desktopItem = makeDesktopItem {
```

Comment:
Is it better to upstream this dekstop item?

Reply:
> Is it better to upstream this dekstop item?

This is not always desirable for many projects. Desktop items are XGD specific IIRC.